### ISSUE TITLE
API will be slow when not "User-Agent" in header

### ISSUE THREAD

**Issue by skanehira** [moby/moby#47439]
### Description

API will be slow when not "User-Agent" in header.

### Reproduce

1. Execute curl without header.
    ```sh
    $ time curl -s --unix-socket /var/run/docker.sock "http://localhost/containers/json?all=true" > /dev/null 


    ========================
    Program : curl -s --unix-socket /var/run/docker.sock  > /dev/null
    CPU     : 0%
    user    : 0.007s
    system  : 0.005s
    total   : 3.146s
    ========================
    ```

2. Execute curl with "User-Agent" header.
    ```sh
    $ time curl -s --unix-socket /var/run/docker.sock -H "User-Agent: Docker-Client/25.0.3" "http://localhost/containers/json?all=true" > /dev/null 


    ========================
    Program : curl -s --unix-socket /var/run/docker.sock -H   > /dev/null
    CPU     : 19%
    user    : 0.012s
    system  : 0.010s
    total   : 0.112s
    ========================

    ```

### Expected behavior

Without header not be slow.

### docker version

```bash
$ docker version                   
Client:
 Cloud integration: v1.0.35+desktop.10
 Version:           25.0.3
 API version:       1.44
 Go version:        go1.21.6
 Git commit:        4debf41
 Built:             Tue Feb  6 21:13:26 2024
 OS/Arch:           darwin/arm64
 Context:           desktop-linux

Server: Docker Desktop 4.27.2 (137060)
 Engine:
  Version:          25.0.3
  API version:      1.44 (minimum version 1.24)
  Go version:       go1.21.6
  Git commit:       f417435e5f6216828dec57958c490c4f8bae4f98
  Built:            Wed Feb  7 00:39:16 2024
  OS/Arch:          linux/arm64
  Experimental:     false
 containerd:
  Version:          1.6.28
  GitCommit:        ae07eda36dd25f8a1b98dfbf587313b99c0190bb
 runc:
  Version:          1.1.12
  GitCommit:        v1.1.12-0-g51d5e94
 docker-init:
  Version:          0.19.0
  GitCommit:        de40ad0
```

### docker info

```bash
$ docker info
Client:
 Version:    25.0.3
 Context:    desktop-linux
 Debug Mode: false
 Plugins:
  buildx: Docker Buildx (Docker Inc.)
    Version:  v0.12.1-desktop.4
    Path:     /Users/skanehira/.docker/cli-plugins/docker-buildx
  compose: Docker Compose (Docker Inc.)
    Version:  v2.24.5-desktop.1
    Path:     /Users/skanehira/.docker/cli-plugins/docker-compose
  debug: Get a shell into any image or container. (Docker Inc.)
    Version:  0.0.24
    Path:     /Users/skanehira/.docker/cli-plugins/docker-debug
  dev: Docker Dev Environments (Docker Inc.)
    Version:  v0.1.0
    Path:     /Users/skanehira/.docker/cli-plugins/docker-dev
  extension: Manages Docker extensions (Docker Inc.)
    Version:  v0.2.21
    Path:     /Users/skanehira/.docker/cli-plugins/docker-extension
  feedback: Provide feedback, right in your terminal! (Docker Inc.)
    Version:  v1.0.4
    Path:     /Users/skanehira/.docker/cli-plugins/docker-feedback
  init: Creates Docker-related starter files for your project (Docker Inc.)
    Version:  v1.0.0
    Path:     /Users/skanehira/.docker/cli-plugins/docker-init
  sbom: View the packaged-based Software Bill Of Materials (SBOM) for an image (Anchore Inc.)
    Version:  0.6.0
    Path:     /Users/skanehira/.docker/cli-plugins/docker-sbom
  scout: Docker Scout (Docker Inc.)
    Version:  v1.4.1
    Path:     /Users/skanehira/.docker/cli-plugins/docker-scout
WARNING: Plugin "/Users/skanehira/.docker/cli-plugins/docker-scan" is not valid: failed to fetch metadata: fork/exec /Users/skanehira/.docker/cli-plugins/docker-scan: no such file or directory

Server:
 Containers: 11
  Running: 6
  Paused: 0
  Stopped: 5
 Images: 22
 Server Version: 25.0.3
 Storage Driver: stargz
  driver-type: io.containerd.snapshotter.v1
 Logging Driver: json-file
 Cgroup Driver: cgroupfs
 Cgroup Version: 2
 Plugins:
  Volume: local
  Network: bridge host ipvlan macvlan null overlay
  Log: awslogs fluentd gcplogs gelf journald json-file local splunk syslog
 Swarm: inactive
 Runtimes: io.containerd.runc.v2 runc
 Default Runtime: runc
 Init Binary: docker-init
 containerd version: ae07eda36dd25f8a1b98dfbf587313b99c0190bb
 runc version: v1.1.12-0-g51d5e94
 init version: de40ad0
 Security Options:
  seccomp
   Profile: unconfined
  cgroupns
 Kernel Version: 6.6.12-linuxkit
 Operating System: Docker Desktop
 OSType: linux
 Architecture: aarch64
 CPUs: 10
 Total Memory: 4.808GiB
 Name: docker-desktop
 ID: de692d8b-bfab-4b8c-ad5f-3b294c40137b
 Docker Root Dir: /var/lib/docker
 Debug Mode: false
 HTTP Proxy: http.docker.internal:3128
 HTTPS Proxy: http.docker.internal:3128
 No Proxy: hubproxy.docker.internal
 Experimental: false
 Insecure Registries:
  hubproxy.docker.internal:5555
  127.0.0.0/8
 Live Restore Enabled: false

WARNING: daemon is not using the default seccomp profile
```

**Comment by thaJeztah** [comment]
Hm.. I'm not seeing this on a Linux machine, so wondering if this is something specific to Docker Desktop (which has a proxy for the API for forwarding requests into the Docker Desktop VM);

```bash
time curl -s --unix-socket /var/run/docker.sock "http://localhost/containers/json?all=true" > /dev/null

real	0m0.013s
user	0m0.005s
sys	0m0.004s

time curl -s --unix-socket /var/run/docker.sock -H "User-Agent: Docker-Client/25.0.3" "http://localhost/containers/json?all=true" > /dev/null

real	0m0.012s
user	0m0.004s
sys	0m0.004s
```

**Comment by thaJeztah** [comment]
Actually, not seeing this on my Docker Desktop as well 🤔 

```bash
time curl -s --unix-socket /var/run/docker.sock "http://localhost/containers/json?all=true" > /dev/null

real	0m0.139s
user	0m0.003s
sys	0m0.006s

time curl -s --unix-socket /var/run/docker.sock -H "User-Agent: Docker-Client/25.0.3" "http://localhost/containers/json?all=true" > /dev/null

real	0m0.137s
user	0m0.003s
sys	0m0.004s
```

**Comment by vvoland** [comment]
@skanehira Can you still reproduce this? Are you sure it's actually related to the `User-Agent` header?

**Comment by skanehira** [comment]
> ```shell
> time curl -s --unix-socket /var/run/docker.sock "http://localhost/containers/json?all=true" > /dev/null
> ```

Yes, this still reproduce

```sh
$ time curl -s --unix-socket /var/run/docker.sock "http://localhost/containers/json?all=true" > /dev/null


========================
Program : curl -s --unix-socket /var/run/docker.sock  > /dev/null
CPU     : 0%
user    : 0.003s
system  : 0.006s
total   : 3.089s
========================

$ time curl -s --unix-socket /var/run/docker.sock -H "User-Agent: Docker-Client/25.0.3" "http://localhost/containers/json?all=true" > /dev/null 


========================
Program : curl -s --unix-socket /var/run/docker.sock -H   > /dev/null
CPU     : 21%
user    : 0.003s
system  : 0.005s
total   : 0.037s
========================
```

docker version

```
$ docker version 
Client:
 Cloud integration: v1.0.35+desktop.13
 Version:           26.0.0
 API version:       1.45
 Go version:        go1.21.8
 Git commit:        2ae903e
 Built:             Wed Mar 20 15:14:46 2024
 OS/Arch:           darwin/arm64
 Context:           desktop-linux

Server: Docker Desktop 4.29.0 (145265)
 Engine:
  Version:          26.0.0
  API version:      1.45 (minimum version 1.24)
  Go version:       go1.21.8
  Git commit:       8b79278
  Built:            Wed Mar 20 15:18:02 2024
  OS/Arch:          linux/arm64
  Experimental:     false
 containerd:
  Version:          1.6.28
  GitCommit:        ae07eda36dd25f8a1b98dfbf587313b99c0190bb
 runc:
  Version:          1.1.12
  GitCommit:        v1.1.12-0-g51d5e94
 docker-init:
  Version:          0.19.0
  GitCommit:        de40ad0
```

docker info

```
$ docker info    
Client:
 Version:    26.0.0
 Context:    desktop-linux
 Debug Mode: false
 Plugins:
  buildx: Docker Buildx (Docker Inc.)
    Version:  v0.13.1-desktop.1
    Path:     /Users/skanehira/.docker/cli-plugins/docker-buildx
  compose: Docker Compose (Docker Inc.)
    Version:  v2.26.1-desktop.1
    Path:     /Users/skanehira/.docker/cli-plugins/docker-compose
  debug: Get a shell into any image or container. (Docker Inc.)
    Version:  0.0.27
    Path:     /Users/skanehira/.docker/cli-plugins/docker-debug
  dev: Docker Dev Environments (Docker Inc.)
    Version:  v0.1.2
    Path:     /Users/skanehira/.docker/cli-plugins/docker-dev
  extension: Manages Docker extensions (Docker Inc.)
    Version:  v0.2.23
    Path:     /Users/skanehira/.docker/cli-plugins/docker-extension
  feedback: Provide feedback, right in your terminal! (Docker Inc.)
    Version:  v1.0.4
    Path:     /Users/skanehira/.docker/cli-plugins/docker-feedback
  init: Creates Docker-related starter files for your project (Docker Inc.)
    Version:  v1.1.0
    Path:     /Users/skanehira/.docker/cli-plugins/docker-init
  sbom: View the packaged-based Software Bill Of Materials (SBOM) for an image (Anchore Inc.)
    Version:  0.6.0
    Path:     /Users/skanehira/.docker/cli-plugins/docker-sbom
  scout: Docker Scout (Docker Inc.)
    Version:  v1.6.3
    Path:     /Users/skanehira/.docker/cli-plugins/docker-scout
WARNING: Plugin "/Users/skanehira/.docker/cli-plugins/docker-compose_bk" is not valid: plugin candidate "compose_bk" did not match "^[a-z][a-z0-9]*$"
WARNING: Plugin "/Users/skanehira/.docker/cli-plugins/docker-scan" is not valid: failed to fetch metadata: fork/exec /Users/skanehira/.docker/cli-plugins/docker-scan: no such file or directory

Server:
 Containers: 19
  Running: 10
  Paused: 0
  Stopped: 9
 Images: 27
 Server Version: 26.0.0
 Storage Driver: overlay2
  Backing Filesystem: extfs
  Supports d_type: true
  Using metacopy: false
  Native Overlay Diff: true
  userxattr: false
 Logging Driver: json-file
 Cgroup Driver: cgroupfs
 Cgroup Version: 2
 Plugins:
  Volume: local
  Network: bridge host ipvlan macvlan null overlay
  Log: awslogs fluentd gcplogs gelf journald json-file local splunk syslog
 Swarm: inactive
 Runtimes: io.containerd.runc.v2 runc
 Default Runtime: runc
 Init Binary: docker-init
 containerd version: ae07eda36dd25f8a1b98dfbf587313b99c0190bb
 runc version: v1.1.12-0-g51d5e94
 init version: de40ad0
 Security Options:
  seccomp
   Profile: unconfined
  cgroupns
 Kernel Version: 6.6.22-linuxkit
 Operating System: Docker Desktop
 OSType: linux
 Architecture: aarch64
 CPUs: 10
 Total Memory: 11.67GiB
 Name: docker-desktop
 ID: 89377bb2-a81c-42ef-98ed-dd5cdf2e05c1
 Docker Root Dir: /var/lib/docker
 Debug Mode: false
 HTTP Proxy: http.docker.internal:3128
 HTTPS Proxy: http.docker.internal:3128
 No Proxy: hubproxy.docker.internal
 Labels:
  com.docker.desktop.address=unix:///Users/skanehira/Library/Containers/com.docker.docker/Data/docker-cli.sock
 Experimental: false
 Insecure Registries:
  hubproxy.docker.internal:5555
  127.0.0.0/8
 Live Restore Enabled: false

WARNING: daemon is not using the default seccomp profile
```

**Comment by vvoland** [comment]
Are you sure it's actually the `User-Agent`? Can you try it in reverse order, just to make sure it's not some FS cache related thing?

**Comment by skanehira** [comment]
> Are you sure it's actually the User-Agent? Can you try it in reverse order, just to make sure it's not some FS cache related thing?

I tried

<img width="1154" alt="image" src="https://github.com/moby/moby/assets/7888591/55a41251-b0e1-46c0-8b2e-ee19bf71b5c5">

**Comment by thaJeztah** [comment]
@skanehira I went looking into some of the code used in Docker Desktop w.r.t. port-mapping (Docker Desktop needs to do some under-the-hood magic to make port-mapping work); is it possible you have some container that failed to start / is in a broken state?

If possible, and if you're still able to reproduce, could you send a diagnostics report and post the ID of the report? https://docs.docker.com/desktop/troubleshoot/overview/#diagnose-from-the-app

**Comment by skanehira** [comment]
> @skanehira I went looking into some of the code used in Docker Desktop w.r.t. port-mapping (Docker Desktop needs to do some under-the-hood magic to make port-mapping work); is it possible you have some container that failed to start / is in a broken state?
> 
> If possible, and if you're still able to reproduce, could you send a diagnostics report and post the ID of the report? https://docs.docker.com/desktop/troubleshoot/overview/#diagnose-from-the-app

Thank you :)
This is ID of report
`578E67C2-E9E7-4166-8E1A-EA573473B406/20240416183205`

**Comment by thaJeztah** [comment]
Thank you!

**Comment by thaJeztah** [comment]
This issue was related to a bug in Docker Desktop; a fix for this was released as part of Docker Desktop 4.30 and up; if you're on an older version of Docker Desktop, updated to 4.30; if you're still experiencing this issue after upgrading, it's best to open a ticket in the Docker Desktop issue tracker (https://github.com/docker/for-mac/issues or https://github.com/docker/for-win/issues)


I'll close this ticket here, because the bug is not in the code maintained in this repository, but feel free to continue the conversation 👍
